### PR TITLE
Add `StringTools.contains`

### PR DESCRIPTION
**Pull request by markknol**
Additions:

* `StringTools.eregReplace`:
   Motivation is that this allows chained replaces, convenience function for `ereg.replace(s, by)`. 
* `StringTools.contains`:
   Motivation is to disallow magic numbers in code, convenience method for `indexOf != -1` with null-check

Contains unit tests and API documentation.

Demo:  https://try.haxe.org/#5b60D

### CODE REVIEW

**Comment by Simn** [comment]
I wouldn't put the full replace documentation there because it's redundant. IMO this should use `@see` or whatever we use in these cases.

Also I really don't like this whole "ereg" naming... But I guess we'll have to live with that.

**Comment by markknol** [comment]
Thanks, I shortened the docs.

As for the naming, we already have a normal replace, so think this second best name?

**Comment by RealyUniqueName** [comment]
Maybe it's better to make them static functions in `EReg` instead of `StringTools`.

**Comment by markknol** [comment]
They are mostly designed to be used as a static extension, and you normally wouldn't do `using EReg` (?)

**Comment by markknol** [comment]
@RealyUniqueName any idea why PHP fails?

**Comment by RealyUniqueName** [comment]
Because `String.indexOf('')` was never tested.
@ncannasse should we follow JS behavior for empty strings in `indexOf('', startIndex)` and `lastIndexOf('', startIndex)`?

**Comment by ncannasse** [comment]
@RealyUniqueName yes :)

**Comment by RealyUniqueName** [comment]
I pushed 8f7c970 with `indexOf('')` and `lastIndexOf('')` specification. That should fix the issue with php.

**Comment by back2dos** [comment]
If there's a chance to allow overloading for inline functions any time soon, then I would suggest not to introduce `eregReplace`.

File: std/StringTools.hx [hunk 1]
```diff
@@ -598,6 +598,25 @@ class StringTools {
...
+		return ereg.replace(str, by);
+	}
+
+	/**
+		Returns `true` if `str` contains `value` and returns `false` otherwise.
+		Returns `false` if `str` is `null`.
+		When `value` is `null`, the result is undefined.
+	**/
+	public static inline function contains(str : String, value : String) : Bool {
+		return if (str == null) false else str.indexOf(value) != -1;
```
Comment (SlavaRa) [inline std/StringTools.hx]: simplifying:
```haxe
return str != null && str.indexOf(value) != -1;
```

File: std/php/_std/StringTools.hx [hunk 2]
```diff
@@ -220,4 +220,11 @@ import php.*;
 		}
 	}
 
+	public static inline function eregReplace(str:String, ereg:EReg, by:String ) : String {
+		return ereg.replace(str, by);
+	}
+
+	public static inline function contains(str:String, value:String) : Bool {
+		return if (str == null) false else str.indexOf(value) != -1;
```
Comment (SlavaRa) [inline std/php/_std/StringTools.hx]: simplifying:
```haxe
return str != null && str.indexOf(value) != -1;
```

File: std/StringTools.hx (hunk 1 above)
Comment (nadako) [reply std/StringTools.hx]: I'm not sure that we even want to specify `str == null` behaviour here, especially given that `StringTools` is meant to be used as a static extension, and I think it's weird that `s.contains(...)` wouldn't fail if `s` is `null`.

But if we do, the type hint should be `Null<String>`.

**Comment by ncannasse** [comment]
Note: I'm not a big fan of having a dependency on regular expressions in StringTools. I think so far they are quite optional when it comes to using the std library.

**Comment by Gama11** [comment]
Could we have `StringTools.contains()` in Haxe 4? :)

It looks like the conensus is to remove `eregReplace()` and the null check from this PR for now, then it'd be mergeable?

**Comment by ncannasse** [comment]
That would work for me

**Comment by Gama11** [comment]
Good to merge? As far as I can tell, the Travis failures are the same ones as on dev.

**Comment by Simn** [comment]
Are we really sure we want this on StringTools? I know I like trolling gama with this, but I find it quite odd that `(s : Dynamic).indexOf("str") != -1` would work while `(s : Dynamic).contains("str")` does not.

I don't usually care about Dynamic, but when it comes to convenience functions I would prefer to not introduce additional Dynamic problems.

**Comment by RealyUniqueName** [comment]
Let's force people to properly type their things by making convenient methods not available for `Dynamic` :)

**Comment by Gama11** [comment]
To me this seems more similar to other convenience functions like `startsWith()` and `endsWith()`, rather than "core functions" like `indexOf()`. I don't really care if this works with Dynamic or not.

Either way, I guess it would not be a breaking change to still support `Dynamic.contains()` later on?

**Comment by Gama11** [comment]
Besides, there's several `Array` methods which don't consistently work with `Dynamic`, and nobody seems to care about that either: https://github.com/HaxeFoundation/haxe/issues/3330

**Comment by Simn** [comment]
Fine, I'm just playing Devil's/Dynamic's Advocate.

**Comment by markknol** [comment]
Thanks!!!

**Pull request merged**